File: 0016.3SumClosest/3SumClosest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        numslen = len(nums)
        closest = 0x7fffffff
        for i in range(numslen - 2):
            if i > 0 and nums[i] == nums[i - 1]: continue
            j = i + 1
            k = numslen - 1
            while j < k:
                three = nums[i] + nums[j] + nums[k]
                dis = abs(three - target)
                if dis < abs(closest - target): closest = three
                if three > target:
                    k -= 1
                elif three < target:
                    j += 1
                else:
                    break
            if j < k:
                break
        return closest
                
        

File: 0016.3SumClosest/3SumClosest.c
int mycompar(const void *a, const void *b)
{
	int *x = (int *)a;
	int *y = (int *)b;
	if (*x > *y) return 1;
	return *x == *y? 0 : -1;
}

int threeSumClosest(int* nums, int numsSize, int target) 
{
	int close = INT_MAX / 2;
	int i, j, k = 0, sum;
	qsort(nums, numsSize, sizeof(int), mycompar);
	
	for (i = 0; i < numsSize - 2; i ++) {
        if (i > 0 && nums[i] == nums[i - 1]) continue;
		j = i + 1;
        k = numsSize - 1;
        while (j < k) {
            sum = nums[i] + nums[j] + nums[k];
            if (abs(close - target) > abs(sum - target))
                close = sum;
            if (sum > target) k --;
            else if (sum < target) j ++;
            else break;
        }
        if (j < k) break;
	}

	return close;
}